sdk/metricsadvisor/ai-metrics-advisor/test/public/dataSourceCred.spec.ts

```diff
@@ -0,0 +1,287 @@
+          tenantId: "updated-tenant",
+          type: "ServicePrincipal"
+        } as ServicePrincipalDatasourceCredential;
+        await client.updateDatasourceCredential(
+          createdServicePrincipalCredId,
+          servicePrincipalCredentialPatch
+        );
+        const updated = await client.getDatasourceCredential(createdServicePrincipalCredId);
+        assert.ok(updated.id, "Expecting valid datasource credential");
+        assert.equal(updated.description, servicePrincipalCredentialPatch.description);
+        assert.equal(updated.type, servicePrincipalCredentialPatch.type);
+        assert.equal(updated.name, servicePrincipalCredentialPatch.name);
+        assert.equal(
+          (updated as ServicePrincipalDatasourceCredential).clientId,
+          servicePrincipalCredentialPatch.clientId
+        );
+      });
+
+      it("creates service principal in keyvault credential", async function() {
+        const servicePrincipalInKVCred: ServicePrincipalInKeyVaultDatasourceCredential = {
+          ...datasourceCredential,
+          name: "ExampleSPinKVCredential",
+          type: "ServicePrincipalInKV",
+          tenantId: "tenant-id",
+          keyVaultEndpoint: "keyvault-endpoint",
+          keyVaultClientId: "keyvault-client-id",
+          keyVaultClientSecret: "keyvault-client-secret",
+          servicePrincipalIdNameInKV: "service-principal-in-kv",
+          servicePrincipalSecretNameInKV: "service-principal-secret-name-in-kv"
+        };
+
+        const createdServicePrincipalInKVCred = await client.createDatasourceCredential(
+          servicePrincipalInKVCred
+        );
+        assert.ok(createdServicePrincipalInKVCred.id, "Expecting valid datasource credential");
+        createdServicePrincipalInKVCredId = createdServicePrincipalInKVCred.id!;
+        assert.equal(createdServicePrincipalInKVCred.name, servicePrincipalInKVCred.name);
+        assert.equal(
+          createdServicePrincipalInKVCred.description,
+          servicePrincipalInKVCred.description
+        );
+        assert.equal(createdServicePrincipalInKVCred.type, servicePrincipalInKVCred.type);
+      });
+
+      it("updates service principal in keyvault credential", async function(this: Context) {
+        if (!createdServicePrincipalInKVCredId) {
+          this.skip();
+        }
+        const servicePrincipalInKVCredentialPatch = {
+          name: "UpdatedSPinKVCred",
+          description: "updated description",
+          keyVaultEndpoint: "updated-keyvault-endpoint",
+          keyVaultClientId: "updated-keyvault-client-id",
+          keyVaultClientSecret: "updated-keyvault-client-secret",
+          servicePrincipalIdNameInKV: "updated-service-principal-in-kv",
+          servicePrincipalSecretNameInKV: "updated-service-principal-secret-name-in-kv",
+          tenantId: "updated-tenant",
+          type: "ServicePrincipalInKV"
+        } as ServicePrincipalInKeyVaultDatasourceCredential;
```

Comment:
We shouldn't cast to `ServicePrincipalInKeyVaultDatasourceCredential`?

Reply:
we changed these to individual patch types and cast to those types